modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/StorageOptions.kt

```diff
@@ -0,0 +1,94 @@
+package au.com.shiftyjelly.pocketcasts.repositories.file
+
+import android.content.Context
+import android.os.Environment
+import au.com.shiftyjelly.pocketcasts.localization.R
+import java.io.File
+
+class StorageOptions {
+
+    private var folderLocations: ArrayList<FolderLocation>? = null
+
+    fun getFolderLocations(context: Context): List<FolderLocation> {
+        if (folderLocations == null) {
+            folderLocations = arrayListOf()
+            val confirmedMountPoints = arrayListOf<String>()
+
+            val externalDirs = context.getExternalFilesDirs(null)
+            if (externalDirs != null && externalDirs.isNotEmpty()) {
+                for (directory in externalDirs) {
+                    if (directory != null) {
+                        confirmedMountPoints.add(directory.absolutePath)
+                    }
+                }
+            }
+
+            testAndCleanMountsList(confirmedMountPoints)
+            determineFolderLocations(confirmedMountPoints, context)
+        }
+
+        return folderLocations!!
+    }
+
+    private fun testAndCleanMountsList(confirmedMountPoints: ArrayList<String>) {
+        /*
+		 * Now that we have a cleaned list of mount paths Test each one to make
+		 * sure it's a valid and available path. If it is not, remove it from
+		 * the list.
+		 */
+        val iterator = confirmedMountPoints.iterator()
+        while (iterator.hasNext()) {
+            val mount = iterator.next()
+            val folder = File(mount)
+            if (!folder.exists() || !folder.isDirectory || !folder.canWrite()) {
+                iterator.remove()
+            }
+        }
+    }
+
+    private fun determineFolderLocations(
+        confirmedMountPoints: ArrayList<String>,
+        context: Context
+    ) {
+        if (confirmedMountPoints.isEmpty()) return
+
+        var externalSDCardCount = 1
+        val firstMountPoint = confirmedMountPoints[0]
+        confirmedMountPoints.removeAt(0)
+
+        val resources = context.resources
+        // the first mount point is different to the rest
+        if (!Environment.isExternalStorageRemovable()
+            || Environment.isExternalStorageEmulated()
+        ) {
+            folderLocations!!.add(
```

Comment:
Nitpick: We might want to avoid using the double-bang(!!) operator inside `determineFolderLocations(...)`. You can add a `requireNotNull(...)` with an explicit message that folder location cannot be null. 

This is just a personal preference and I don't feel strongly about it.

Reply:
Hi, i see that explicit message is helpful. Pushed the new one